src/navigation/tabs/tabs.jsx

```diff
@@ -1,86 +1,674 @@
+        this.onResize = this.onResize.bind(this);
+        this.onResizeThrottled = throttle(this.onResize, props.resizeThrottle, { trailing: true });
+        this.selectedTabKeyProp = props.selectedTabKey;
+        this.setTabsDimensions = this.setTabsDimensions.bind(this);
+        this.tabRefs = {};
+    }
+
+    componentDidMount() {
+        this.setTabsDimensions();
+    }
+
+    shouldComponentUpdate(nextProps, nextState) {
+        const {
+            items,
+        } = this.props;
+
+        const {
+            blockWidth,
+            selectedTabKey,
+            tabsTotalWidth,
+            items: itemsState,
+        } = this.state;
+
+        const shouldUpdate = (
+            !isEqual(nextProps.items, items) ||
+            nextState.tabsTotalWidth !== tabsTotalWidth ||
+            nextState.blockWidth !== blockWidth ||
+            nextProps.selectedTabKey !== this.selectedTabKeyProp ||
+            nextState.selectedTabKey !== selectedTabKey ||
+            !isEqual(itemsState, nextState.items)
+        );
+
+        return shouldUpdate;
     }
 
     componentDidUpdate(prevProps) {
-        if (this.props.selected !== prevProps.selected) {
-            this.setState({ selected: this.props.selected });
+        const {
+            items,
+            selectedTabKey,
+        } = this.props;
+
+        const {
+            items: prevItems,
+        } = prevProps;
+
+        const didPropsItemsChange = !isEqual(prevItems, items);
+        const didSelectedTabKeyChange = this.selectedTabKeyProp !== selectedTabKey;
+        const shouldUpdateState = didPropsItemsChange || didSelectedTabKeyChange;
+
+        if (shouldUpdateState) {
+            const newState = {
+                ...(didSelectedTabKeyChange && {
+                    selectedTabKey,
+                }),
+                ...(didPropsItemsChange && {
+                    items,
+                    selectedTabKey,
+                }),
+            };
+
+            this.setState(newState);
+        }
+
+        if (didPropsItemsChange) {
+            this.setTabsDimensions();
         }
+
+        this.selectedTabKeyProp = selectedTabKey;
     }
 
-    render() {
+    onResize() {
+        if (this.tabsWrapper) {
+            this.setState({ blockWidth: this.tabsWrapper.offsetWidth });
+        }
+    }
+
+    onChangeTab(nextTabKey, evt) {
         const {
-            children,
-            id,
+            beforeChange,
+            onChange,
         } = this.props;
-        const containerClasses = ClassNames('ui', 'tabs', this.props.className, {
-            'tabs-nest': this.props.nest,
+
+        const {
+            selectedTabKey,
+        } = this.state;
+
+        if (isFunction(beforeChange)) {
+            const beforeChangeRes = beforeChange({ selectedTabKey, nextTabKey });
+
+            if (beforeChangeRes === false) {
+                if (evt) {
+                    evt.preventDefault();
+                }
+
+                return null;
```

Comment:
Can we just use `return;` here (and on Line 262) instead of `return null;`.  I don't think it is necessary for event handler methods to have a return type; they should have `void` return type, right?

Reply:
Fixed.

Reply:
Lookin' better.  Resolving.